### PR TITLE
(FACT-2054) facter execution execute expands shell builtins

### PR DESCRIPTION
**Pull request by GeorgeMrejea**

### CODE REVIEW

File: lib/src/ruby/module.cc [hunk 1]
```diff
@@ -883,7 +883,16 @@ namespace facter { namespace ruby {
                 raise = true;
                 fail_option = ruby.nil_value();
             }
-            return execute_command(ruby.to_string(argv[0]), fail_option, raise, timeout);
+            bool expand = true;
+            try{
+            volatile VALUE expand_value =  ruby.to_symbol("expand");
+            volatile VALUE expand_option = ruby.rb_hash_lookup3(argv[1], ruby.nil_value(), expand_value);
```
Comment (GeorgeMrejea) [inline lib/src/ruby/module.cc]: Maybe not the best way to pass new expand parameter, but leatherman does not compile with this option

File: lib/inc/internal/util/agent.hpp [hunk 2]
```diff
@@ -16,9 +16,20 @@ namespace facter { namespace util { namespace agent {
      * @param exe the name of the executable
      * @return the path to the executable
      */
-    inline std::string which(const std::string& exe) {
+inline std::string which(const std::string& exe) {
```
Comment (ciprianbadescu) [inline lib/inc/internal/util/agent.hpp]: indentation

File: lib/inc/internal/util/agent.hpp [hunk 3]
```diff
@@ -16,9 +16,20 @@ namespace facter { namespace util { namespace agent {
...
 #ifdef FACTER_PATH
-        std::string fixed = leatherman::execution::which(exe, {FACTER_PATH});
+    std::string fixed = leatherman::execution::which(exe, {FACTER_PATH}, true);
+    if (!fixed.empty()) {
+        return fixed;
+    }
+    LOG_WARNING("{1} not found at configured location {2}, using PATH instead", exe, FACTER_PATH);
```
Comment (ciprianbadescu) [inline lib/inc/internal/util/agent.hpp]: I think this check and logging is redundant to the one in which with 3 parameters.

File: lib/src/ruby/module.cc [hunk 4]
```diff
@@ -908,16 +918,19 @@ namespace facter { namespace ruby {
...
     {
         auto const& ruby = api::instance();
-
-        // Expand the command
-        auto expanded = expand_command(command);
-
+        std::string expanded;
+        try {
+        // Expand the command only if expand is true,
+        expanded = leatherman::execution::expand_command(command, leatherman::util::environment::search_paths(), expand);
```
Comment (ciprianbadescu) [inline lib/src/ruby/module.cc]: leatherman::execution is already imported, no need to specify it explicitly here and in similar places

File: lib/tests/ruby/ruby.cc [hunk 5]
```diff
@@ -497,6 +497,16 @@ SCENARIO("custom facts written in Ruby") {
             REQUIRE(re_search(output, boost::regex("ERROR puppetlabs\\.facter - .* command timed out after 1 seconds")));
         }
     }
+    GIVEN("a fact resolution that uses Facter::Core::Execution#execute with expand value set to false") {
```
Comment (ciprianbadescu) [inline lib/tests/ruby/ruby.cc]: We need tests for all combinations of commend/builtin, expand = true/false

File: lib/src/ruby/module.cc [hunk 6]
```diff
@@ -23,6 +23,9 @@
 #include <facter/facts/map_value.hpp>
 #include <internal/ruby/ruby_value.hpp>
 
+#include <iostream>
+#include <typeinfo>
+
```
Comment (ciprianbadescu) [inline lib/src/ruby/module.cc]: not needed, pls remove

File: lib/inc/internal/util/agent.hpp (hunk 2 above)
Comment (GeorgeMrejea) [reply lib/inc/internal/util/agent.hpp]: I have discarded all modification for this file. You are reviewing the version a02a661 which is three years old.

File: lib/inc/internal/util/agent.hpp (hunk 3 above)
Comment (GeorgeMrejea) [reply lib/inc/internal/util/agent.hpp]: See response above

File: lib/src/ruby/module.cc [hunk 7]
```diff
@@ -908,13 +917,16 @@ namespace facter { namespace ruby {
...
         auto const& ruby = api::instance();
-
-        // Expand the command
-        auto expanded = expand_command(command);
-
+        std::string expanded;
+        try {
+        // Expand the command only if expand is true,
+        expanded = expand_command(command, leatherman::util::environment::search_paths(), expand);
+        } catch (const std::exception& e ) {
```
Comment (GabrielNagy) [inline lib/src/ruby/module.cc]: What's the exception we can get here? Aren't exceptions treated already in expand_command? Also we shouldn't use `std::cout` to log them.

**Comment by GeorgeMrejea** [comment]
Done

**Comment by GabrielNagy** [comment]
looks good :+1:, please squash your commits

**Comment by GeorgeMrejea** [comment]
squashed, ready to merge

File: lib/src/ruby/module.cc [hunk 8]
```diff
@@ -908,13 +915,13 @@ namespace facter { namespace ruby {
...
 
-    VALUE module::execute_command(std::string const& command, VALUE failure_default, bool raise, uint32_t timeout)
+    VALUE module::execute_command(std::string const& command, VALUE failure_default, bool raise, uint32_t timeout, bool expand)
     {
         auto const& ruby = api::instance();
-
-        // Expand the command
-        auto expanded = expand_command(command);
-
+
```
Comment (gimmyxd) [inline lib/src/ruby/module.cc]: cpp lint fails on this

File: lib/src/ruby/module.cc [hunk 9]
```diff
@@ -908,13 +915,13 @@ namespace facter { namespace ruby {
...
     {
         auto const& ruby = api::instance();
-
-        // Expand the command
-        auto expanded = expand_command(command);
-
+        
+        // Expand the command only if expand is true,
+        auto expanded = expand_command(command, leatherman::util::environment::search_paths(), expand);
+
```
Comment (gimmyxd) [inline lib/src/ruby/module.cc]: cpp lint fails on this

**Comment by gimmyxd** [comment]
@GeorgeMrejea please rebase with 3.11.x to get https://github.com/puppetlabs/facter/pull/1849/commits/043e08df1c63297d18776be974bdadae1ec9797f

**Pull request merged**